### PR TITLE
 Added route to encrypt request

### PR DESCRIPTION
**Pull request by mrrovot**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -116,82 +131,119 @@ app.use(Helmet())
...
-    logHTTPError(req, 'Invalid address supplied')
-    return res.status(400).send()
-  }
 
-  /* Verify that the caller supplied us with an acceptable callback
+    try {
+      cryptoUtils.decodeAddress(address)
+      walletRequest.address = address
+    } catch (e) {
+      return reject('Invalid address supplied')
```
Comment (brandonlehmann) [inline index.js]: reject should return an error. ie.

`reject(new Error('Invalid address supplied'))`

File: index.js [hunk 2]
```diff
@@ -116,82 +131,119 @@ app.use(Helmet())
...
+
+    /* Verify that the caller supplied us with an acceptable callback
      URL that we'll post back to later */
-  if (callback) {
-    if (callback.substring(0, 4).toLowerCase() !== 'http') {
-      logHTTPError(req, 'Invalid callback URL supplied')
-      return res.status(400).send()
+    if (callback) {
+      if (callback.substring(0, 4).toLowerCase() !== 'http') {
+        return reject('Invalid callback URL supplied')
```
Comment (brandonlehmann) [inline index.js]: reject should return an error. ie.

`reject(new Error('Invalid address supplied'))`

File: index.js [hunk 3]
```diff
@@ -116,82 +131,119 @@ app.use(Helmet())
...
-  /* Verify that the caller has supplied a valid amount to request */
-  if (!atomicAmount || atomicAmount === 0 || atomicAmount < 0) {
-    logHTTPError(req, 'Invalid amount requested')
-    return res.status(400).send()
-  }
+    walletRequest.callback = callback
+
+    /* Verify that the caller has supplied a valid amount to request */
+    if (!atomicAmount || atomicAmount === 0 || atomicAmount < 0) {
+      return reject('Invalid amount requested')
```
Comment (brandonlehmann) [inline index.js]: reject should return an error. ie.

`reject(new Error('Invalid address supplied'))`

File: index.js [hunk 4]
```diff
@@ -116,82 +131,119 @@ app.use(Helmet())
...
-    /* If the caller requested 0 or less or more confirmations than we
+    var requestConfirmations
+    if (confirmations !== false) {
+      /* If the caller requested 0 or less or more confirmations than we
        allow, we're going to reject their request */
-    if (confirmations < 0 || confirmations > Config.maximumConfirmations) {
-      logHTTPError(req, 'Invalid confirmations requested')
-      return res.status(400).send()
+      if (confirmations < 0 || confirmations > Config.maximumConfirmations) {
+        return reject('Invalid confirmations requested')
```
Comment (brandonlehmann) [inline index.js]: reject should return an error. ie.

`reject(new Error('Invalid address supplied'))`

File: index.js [hunk 5]
```diff
@@ -116,82 +131,119 @@ app.use(Helmet())
...
+  return new Promise(function (resolve, reject) {
+    let walletRequest = {}
+
+    if (!atomicAmount) {
+      return reject('Invalid amount supplied')
+    }
 
-  const amount = (atomicAmount / Math.pow(10, Config.coinDecimals))
-  var cancelTimer
+    const amount = atomicAmount / Math.pow(10, Config.coinDecimals)
```
Comment (brandonlehmann) [inline index.js]: the `amount` variable in unused within this scope -- it needs to exist in the context in which it was called

File: index.js [hunk 6]
```diff
@@ -116,82 +131,119 @@ app.use(Helmet())
...
+  })
+}
+
+/* This is the meat and potatoes entry method for the public API
+   aka, submitting a new request for funds to the processing engine */
+app.post('/v1/new', async function (req, res) {
+  const atomicAmount = toNumber(req.body.amount)
+  const callback = req.body.callback || false
+  const address = req.body.address || false
+  const name = req.body.name || false
```
Comment (brandonlehmann) [inline index.js]: `name` is unused within this scope -- I think we lost track of it somewhere

File: index.js [hunk 7]
```diff
@@ -116,82 +131,119 @@ app.use(Helmet())
...
+  const callerData = req.body.userDefined || {}
+  const confirmations = toNumber(req.body.confirmations)
+  const hashedRequest = req.body.hashedRequest || false
+
+  var cancelTimer
+  let walletRequest
+
+  if (hashedRequest) {
+    const crypto = new AES({ password: privateRabbitEncryptionKey })
+    const decryptedRequest = await crypto.decrypt(hashedRequest)
```
Comment (brandonlehmann) [inline index.js]: Where is the failure to decrypt the data handled?

File: index.js [hunk 8]
```diff
@@ -16,10 +16,13 @@ const Express = require('express')
 const walletQueue = 'request.wallet'
 const RabbitMQ = require('amqplib')
 const UUID = require('uuid/v4')
+const AES = require('./lib/aes.js')
 
 const publicRabbitHost = process.env.RABBIT_PUBLIC_SERVER || 'localhost'
 const publicRabbitUsername = process.env.RABBIT_PUBLIC_USERNAME || ''
 const publicRabbitPassword = process.env.RABBIT_PUBLIC_PASSWORD || ''
+const privateRabbitEncryptionKey =
```
Comment (brandonlehmann) [inline index.js]: If we aren't using this with rabbit payloads (which we aren't) then the variable name shouldn't be named as such :)

File: index.js [hunk 9]
```diff
@@ -116,82 +131,119 @@ app.use(Helmet())
...
+  const name = req.body.name || false
+  const callerData = req.body.userDefined || {}
+  const confirmations = toNumber(req.body.confirmations)
+  const hashedRequest = req.body.hashedRequest || false
+
+  var cancelTimer
+  let walletRequest
+
+  if (hashedRequest) {
+    const crypto = new AES({ password: privateRabbitEncryptionKey })
```
Comment (brandonlehmann) [inline index.js]: Please move this up to the top of the file -- we shouldn't have to keep re initializing this instance as the password remains unchanged.

File: index.js [hunk 10]
```diff
@@ -247,6 +309,31 @@ app.post('/v1/new', async function (req, res) {
...
+      callback,
+      address,
+      confirmations,
+      callerData
+    )
+    const crypto = new AES({ password: privateRabbitEncryptionKey })
+    const encryptRequest = await crypto.encrypt(walletRequest)
+    res.status(200).json(encryptRequest)
+  } catch (e) {
+    res.status(400).send(e)
```
Comment (brandonlehmann) [inline index.js]: this needs to return and same with the one two lines above

File: index.js (hunk 7 above)
Comment (mrrovot) [reply index.js]: Should I add a try catch here?

Comment (brandonlehmann) [reply index.js]: That would do it but tbh I cringe a little inside whenever i see try/catch unless unavoidable. In any event, it should be returning a status code of 400 if that data exists but it doesn't decrypt

Comment (mrrovot) [reply index.js]: I tested and all works besides this check, I have a hard time finding a proper way to handle this failure in ae.js

Comment (brandonlehmann) [reply index.js]: Can you push up your additional changes?

Comment (mrrovot) [reply index.js]: https://github.com/TurtlePay/turtlepay-public-api/tree/5d300a5045a6a6680925b6e6e02c16b5d36adef8

Comment (mrrovot) [reply index.js]: this would be latest: 
https://github.com/TurtlePay/turtlepay-public-api/blob/e751d84b0004cc259145d2048100d682dfa6f131/index.js

**Comment by brandonlehmann** [comment]
Incorporated into https://github.com/TurtlePay/turtlepay-public-api/commit/61b1042e389318c698917fae0c58b205974929ed

**Pull request closed without merging**